[DR-3027] Optionally turn off pushing events to mixPanel

samanehsan opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3027

The Data Repo is sending too many events to MixPanel, so this PR adds the ability to turn off pushing events to mixPanel if a Bard event has the `pushToMixPanel` property set to `false`. If the property is not defined, Bard will send the event to MixPanel so this change does not impact existing services that use Bard.

Another change is that now all Bard event logs are sent to bigQuery (so even if a service opts out of mixPanel it is still possible to query the event logs). This was done by manually creating a log sink for dev and prod:
- Dev: The [bq-metrics](https://console.cloud.google.com/logs/router?project=terra-bard-dev) sink sends logs with `logName="projects/terra-bard-dev/logs/metrics"` to this [warehouse](console.cloud.google.com/bigquery?ws=!1m5!1m4!4m3!1sbroad-dsde-analytics-dev!2swarehouse!3smetrics) dataset (it looks like the automatically created table is called "metrics").
- Prod: The [bq-metrics](https://console.cloud.google.com/logs/router?authuser=3&project=terra-bard-prod) sink sends logs with `logName="projects/terra-bard-prod/logs/metrics"` to this [warehouse](console.cloud.google.com/bigquery?ws=!1m5!1m4!4m3!1sbroad-dsde-prod-analytics-dev!2swarehouse!3smetrics) dataset

**Note:** GCP creates a service account for each log sink that writes to the bigQuery dataset. I granted the service accounts bigQuery.dataEditor permissions on the destination datasets following: https://cloud.google.com/logging/docs/export/configure_export_v2#destination-paths

**Testing:**
1. Running Bard
- Follow the README instructions to run Bard locally. Change the port so that it does not conflict with TDR:
`GCP_PROJECT=terra-bard-dev GOOGLE_APPLICATION_CREDENTIALS=<path-to-key-file> PORT=8081 yarn run start-dev`

2. Running TDR
- Run this TDR branch locally https://github.com/DataBiosphere/jade-data-repo/pull/1471 and set the [bard base path](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/resources/application.properties#L124) to `http://localhost:8081`.
- Send a request to one of the GET endpoints
- Check the BigQuery table for the event (need to be logged in as your firecloud.org user). Note there can be a delay in BigQuery recording the `_PARTITIONTIME`.

samanehsan committed:
Add bigquery package

samanehsan committed:
Write mixpanel events to bigquery

TODO: Use the data warehouse dataset

samanehsan committed:
Use bigquery warehouse dataset

Define the google project, bq
dataset and table to use in the
bard env configs.

samanehsan committed:
Fix adding bigquery library

samanehsan commented on config/alpha.json:
```diff
@@ -1,5 +1,8 @@
-  "samRoot": "https://sam.dsde-alpha.broadinstitute.org"
+  "samRoot": "https://sam.dsde-alpha.broadinstitute.org",
+  "bigQueryProject": "broad-dsde-prod-analytics-dev",
+  "bigQueryDatasetId": "warehouse",
```
I'm not sure if this is the right dataset for alpha to use (same with staging and perf)

nmalfroy replied:
One thing that we slacked about: should we maybe take the approach of logging and setting up a sink?  I'm just worried that this will fail writing to BQ when there's more traffic

samanehsan replied:
I'll look into that! I think it could simplify the implementation here too

samanehsan committed:
Fix checking useBigQuery property

nmalfroy commented on config/staging.json:
```diff
@@ -1,5 +1,8 @@
-  "samRoot": "https://sam.dsde-staging.broadinstitute.org"
+  "samRoot": "https://sam.dsde-staging.broadinstitute.org",
+  "bigQueryProject": "broad-dsde-prod-analytics-dev",
+  "bigQueryDatasetId": "warehouse",
```
I think this should be the scratch one to start.  Mind confirming with Alex?

nmalfroy replied:
This _might_ be moot if we go the logging approach.

samanehsan replied:
Ah so we still need to define which bigQuery dataset to send the logs to

samanehsan committed:
Use indirect logging sink approach

Instead of explicitly inserting rows
into a bigQuery table, log the event
and use a logging sink to send the
event logs to bigQuery.

nmalfroy reviewed (approved):
Looks good!  Thanks for making the changes and walking through this with me yesterday. Is the sink recorded anywhere or was that just created manually?

okotsopoulos commented on src/server.js:
```diff
@@ -131,7 +131,8 @@ const main = async () => {
    * @api {post} /api/event Log a user event
-   * @apiDescription Records the event to a log and forwards it to mixpanel. Optionally takes an authorization token which must be verified with Sam
+   * @apiDescription Records the event to a log and forwards it to mixpanel. Optionally takes an authorization token which must be verified with Sam.
+   *                 If properties['useBigQuery'] is true, only log the event (the logs will get sent to BigQuery via a log sink instead).
```
What this documentation makes me think is that if `useBigQuery` is true, then we will automatically get the log sent to BigQuery with no additional effort required.  But as far as I understand it, you had to set up log sink(s) for TDR logs.

1. What this property really enables is whether to push data to Mixpanel or not.  Would it be feasible to rework the property name to indicate the behavior driven in this method?  I could see a property `pushToMixpanel` that defaults to true if undefined.
2. Who is the eventual consumer / beneficiary of these logs?  How will they figure out where to look for them in BigQuery, is there any record or hard link to log sinks from our code?

samanehsan replied:
Thanks for the feedback! I reworked the PR to use a `pushToMixpanel` property. I'm not sure who the intended consumer is of these logs actually, since I've only ever used mixPanel 🤔 I can update the README to record the new functionality!

okotsopoulos replied:
Awesome, thanks! The README helps :)

okotsopoulos replied:
Actually, sorry, can you update the docstring as well?  I'd also call out that `pushToMixpanel` defaults to true if unspecified.

samanehsan replied:
Ah thanks for catching that!

okotsopoulos reviewed (commented):
This looks like it will work, but I wanted to raise a concern!

samanehsan committed:
Make pushing to mixpanel opt-out

Instead of opting into using bigQuery,
skip push to mixpanel if the "pushToMixpanel" property
is explicitly set to false.

samanehsan committed:
Update README
